Content.Server/Nyanotrasen/Research/Oracle/OracleSystem.cs

```diff
@@ -1,258 +1,288 @@
             {
-                component.RejectAccumulator = 0;
-                _chat.TrySendInGameICMessage(uid, _random.Pick(component.RejectMessages), InGameICChatType.Speak, true);
+                oracle.Comp.NextRejectTime = _timing.CurTime + oracle.Comp.RejectDelay;
+                _chat.TrySendInGameICMessage(oracle, _random.Pick(oracle.Comp.RejectMessages), InGameICChatType.Speak, true);
             }
+
             return;
         }
 
-        EntityManager.QueueDeleteEntity(args.Used);
-
-        EntityManager.SpawnEntity("ResearchDisk5000", Transform(args.User).Coordinates);
-
-        DispenseLiquidReward(uid, component);
+        DispenseRewards(oracle, Transform(args.User).Coordinates);
+        QueueDel(args.Used);
 
-        var i = _random.Next(1, 4);
+        if (updateRequest)
+            NextItem(oracle);
+    }
 
-        while (i != 0)
-        {
-            EntityManager.SpawnEntity("MaterialBluespace1", Transform(args.User).Coordinates);
-            i--;
-        }
+    private void SendTelepathicInfo(Entity<OracleComponent> oracle, INetChannel client, string message)
+    {
+        var messageWrap = Loc.GetString("chat-manager-send-telepathic-chat-wrap-message",
+            ("telepathicChannelName", Loc.GetString("chat-manager-telepathic-channel-name")),
+            ("message", message));
 
-        if (nextItem)
-            NextItem(component);
+        _chatMan.ChatMessageToOne(ChatChannel.Telepathic,
+            message, messageWrap, oracle, false, client, Color.PaleVioletRed);
     }
 
-    private bool CheckValidity(EntityPrototype given, EntityPrototype target)
+    private bool IsCorrectItem(EntityPrototype given, EntityPrototype target)
     {
-        // 1: directly compare Names
-        // name instead of ID because the oracle asks for them by name
-        // this could potentially lead to like, labeller exploits maybe but so far only mob names can be fully player-set.
+        // Nyano, what is this shit?
+        // Why are we comparing by name instead of prototype id?
+        // Why is this ever necessary?
+        // What were you trying to accomplish?!
         if (given.Name == target.Name)
             return true;
 
         return false;
     }
 
-    private void DispenseLiquidReward(EntityUid uid, OracleComponent component)
+    private void DispenseRewards(Entity<OracleComponent> oracle, EntityCoordinates throwTarget)
     {
-        if (!_solutionSystem.TryGetSolution(uid, OracleComponent.SolutionName, out var fountainSol))
-            return;
+        foreach (var rewardRandom in oracle.Comp.RewardEntities)
+        {
+            // Spawn each reward next to oracle and throw towards the target
+            var rewardProto = _protoMan.Index(rewardRandom).Pick(_random);
+            var reward = EntityManager.SpawnNextToOrDrop(rewardProto, oracle);
+            _throwing.TryThrow(reward, throwTarget, recoil: false);
+        }
```

Comment:
Same here, Iirc research tier is an integer? You could for instance make it drop one item per integer value of the research tier(or 1, if there was no research tier).